docs/refman/ppad.8.pod
#
# mouse:~ppr/src/docs/refman/ppad.8.pod
# Copyright 1995--2012, Trinity College Computing Center.
# Written by David Chappell.
#
# This file is part of PPR.  You can redistribute it and modify it under the
# terms of the revised BSD licence (without the advertising clause) as
# described in the accompanying file LICENSE.txt.
#
# Last modified 25 May 2012.
#

=head1 NAME

ppad - PPR administrator's utility

=head1 SYNOPSIS

B<ppad> [B<--help>] [B<--version>] [B<-M>] [B<--user> I<username>] [B<-d> I<debuglevel>] [I<subcommand>] [I<parameters> ...]

=head1 DESCRIPTION

B<Ppad> is the utility used to manage PPR, that is, to create and delete
printers and groups, to set their parameters, and to define known types of
media.

The B<--version> switch prints the PPR version number.

The B<--help> switch prints abreviated instructions for use and direction for
obtaining additional help.

The B<-d> option can be used to turn on debugging messages which describe what
B<ppad> is doing.  The default level is 0.  Using B<-d 1> will cause B<ppad> to
tell what printer and group configuration files it is examining and editing.
Higher levels display more information.  A level 2 and above B<ppad> will
display information about what it learns from PPD files and
F</usr/share/ppr/lib/mfmodes.conf>.  At level 3 and above it will display every
mode line in F<mfmodes.conf> and every line read from or written to the printer
or group configuration file.  Level 6 and above shows most of the lines in
every PPD file that is read.

Most operations performed by this command require PPR administrator privileges.
Only those operations which make no changes to the spooler configuration may be
performed by non-privileged users.  This command defines a privileged user as
the users "ppr" and "pprwww", a user with a UID of 0 (root), or a user listed
in the file F</etc/ppr/acl/ppad.allow>.

B<Ppad> has a number of subcommands.  Each subcommand has zero or more
parameters.  The subcommands are described below.

If the B<ppad> command is invoked without sub-command arguments it will enter
interactive mode.  In interactive mode it reads commands from stdin and
executes them.

If the B<-M> switch is used, B<ppad> will adopt a terser style of output, more
suitable to parsing by other programs.

If the B<--user> switch is used, B<ppad> will behave as if invoked by the
indicated user.


=head2 Printer Management Subcommands

=over 4

=item B<ppad show> I<printer>

Displays the configuration of the indicated printer.  The option selections
described by the output of this command may be changed using the commands
described below.

=item B<ppad copy> I<printer> I<new_printer>

Copy I<printer> to I<new_printer>.  It is sometimes difficult to determine the
correct settings for a particular model of printer.  This command allows you
to create a new printer queue by copying a working one and changing a few
of the settings.

=item B<ppad comment> [I<printer comment>]

Sets the comment field of the indicated printer.  Highly recommended!  This
comment field is often displayed in user interfaces which allow the user
to choose a printer.  You should set this field to something that will help
the user to identify the printer.  If you really want to clear the comment
field, omit the value.

=item B<ppad location> [I<printer location>]

This command can be used to attach a location description to the printer
queue.  This information could be useful to field service personel who want
to visit the printer.  If the value is omited, the location description 
field will be cleared.

=item B<ppad department> [I<printer department>]

This command can be used to indicate the department of your organization
which uses the printer.  This information is for the edification of human
operators and field service personel.  If the value is omited, the department
field will be cleared.

=item B<ppad contact> [I<printer contact>]

This command provides a way to specify the name and telephone number of a
person that a human operator may want to contact when he notices a problem
with the printer.  This should be a person who works near the printer.  
If the value is omited, the printer contact field will be cleared.

=item B<ppad interface> I<printer interface address>

specifies the means by which PPR will send the jobs to the printer and the
physical address of the printer.  If the print queue I<printer> does not
already exist, it is created.  The I<interface> is the name of a program in
F</usr/lib/ppr/interfaces> which can open a connection to the printer, perhaps
over the network or over a serial cable.  The I<address> is a parameter
which is passed to the interface program.  The precise meaning and format of
the I<address> parameter depend on the interface program selected.  It might
be the device name of a serial port, or the AppleTalk name of the printer.

The interfaces supplied with the current distribution are
B<simple>,
B<serial>,
B<parallel>,
B<dummy>,
B<atalk>,
B<tcpip>,
B<lpr>,
B<clispool>,
B<smb>,
B<gssimple>,
B<gsserial>,
B<gsparallel>,
B<gsatalk>.
B<gstcpip>,
B<gslpr>,
and B<gssmb>.

These interfaces are described in the B<ppad-interfaces(8)> manpage.

=item B<ppad options> I<printer> I<options>

sets the options string which will be passed to the interface program.  Most
interface programs ignore this parameter.  The options string is a sequence of
space separated I<name>=I<value> pairs.  The I<name>=I<value> pair recognized
by each interface supplied with PPR are described in the B<ppad interface>
section of this man page.

Here is a sample "ppad options" command.  This one sets the options for a
printer with a serial interface:

C<$ ppad options myprn 'speed=19200 bits=7 parity=even'>

The available options depend on which interface was selected.  See the
appropriate section in L<ppad-interfaces.8> to learn what
options are available for any given interface.

The options can be removed either by specifying an empty string or omitting the options
string entirely.

=item B<ppad jobbreak I<printer> {none, signal, signal/pjl, control-d, pjl, save/restore, newinterface, default}>>

Specifies the method used to indicate the end of one job and the start of
another to the interface program.  A printer considers a banner page
and the print job proper to be two separate print jobs.  PPR knows
reasonable default job break methods for the standard set of interfaces,
but if you supply your own interface you may have to set this parameter.

The possible settings are as follows:

=over 4

=item B<default>

If you set this mode, then the actual jobbreak method is determined each
time a job is printed by consulting PPR's internal table and the printer's
PPD file.  You can use the B<ppad show> command to determine the actual
jobbreak method decided on.

=item B<control-d>

In this mode a control-D character is sent at the start and end of the job
in order to separate it from other jobs.  This is the default for many
of the interfaces supplied with PPR.

=item B<pjl>

Send control-d's and HP PJL commands to indicate job breaks and to
set the display message.  For many of the interfaces supplied with PPR
this will be the default if the PPD file indicates that the printer
supports PJL.

=item B<signal>

The default for the interface B<atalk>, this method uses a call and response
protocol involving SIGUSR1.  This is the default for the B<atalk> interface
if the PPD file does not indicate that the printer supports PJL.

=item B<signal/pjl>

This selects the SIGUSR1 protocol with the addition of certain HP PJL
commands which do such things as set the printer display panel message to
the name of the user.  This will be the default for the B<atalk> interface
if the PPD file indicates that the printer supports PJL.

=item B<newinterface>

Launches a new copy of the interface program for each segment of the job.
This is the default for the interfaces whose names start with "gs".  For them,
it is also the only setting that makes sense.

=item B<none>

Do not indicate job breaks.  Doing this will make flag pages and multiple copies
produced by certain methods very unreliable.  Do not use this except for
testing.

=item B<save/restore>

Enclose each section of the job in save/restore.  This is only a little
better than B<none>.  Do not use this except for testing.

=back

To return to the default job break method, set it to "default".

=item B<ppad feedback> I<printer> {B<true>, B<False>, B<default>}

Indicates whether the interface program is capable of returning messages
from the printer.  Sometimes this is not possible, as in the case of a
unidirectional interface.  PPR has default feedback settings for the
standard set of interfaces.  To return to the default setting, set it
to B<default>.

=item B<ppad codes> I<printer> {B<Clean7Bit>, B<Clean8Bit>, B<Binary>, B<TBCP>, B<UNKNOWN>, B<default>}

This setting indicates which subset of the 256 possible byte values the
communications chanel will pass through to the printer's PostScript
interpreter.

For any given printer interface program and PPD file there is a default
setting for B<ppad codes>.  To revert to the default setting for the
interface and PPD file, set use the command B<ppad codes> I<printer> B<default>.

The B<ppad codes> setting is new in PPR version 1.40.  Setting it to B<UNKNOWN>
causes PPR to behave as it did in previous versions.  If the setting is
anything other than B<UNKNOWN> then PPR will not attempt to print PostScript
files which claim to require a more extensive byte value set.  For example, if
a printer's codes setting is B<Clean8Bit> and the print job contains a
DSC comment which says "%%DocumentData: Binary" then PPR will not attempt to
print the job on that printer.  Hopefully the job was submitted to a group
and some other member will have a less restrictive codes setting.

Here are the possible settings:

=over 4

=item B<Clean7Bit>

The PostScript code may contain only the characters 0x1B through 0x7E as well
as the three charters 0x0A, 0x0D, and 0x09.  Whenever the last three
appear they must represent line-feed, carriage-return, and tab respectively.

=item B<Clean8Bit>

The same as B<Clean7Bit> except that the characters 0x80 through 0xFF are
also allowed.  This the the default for interfaces including B<serial>,
B<parallel> and B<tcpip>.  These interfaces do not meet the requirements
of B<Binary> because certain code values (such as control-D are used
for control functions.

=item B<Binary>

Any byte value is passed through to the PostScript interpreter.  This seems
to be true of all PostScript printers connected over AppleTalk.  It is also
true of interfaces which run Ghostscript.

=item B<TBCP>

This setting informs PPR that any byte value may be passed through to the
PostScript interpreter as long as control characters are encoded using the
Tagged Binary Communications Protocol.  TBCP is useful when the
communications chanel reserves certain byte values for control purpose.
Often these include control-s, control-q, control-d, and control-t.  With
TBCP it is possible to encode these control codes as two-byte escape sequences.
In the printer, the encoding is undone just before the data is passed to
the PostScript interpreter.

=item B<UNKNOWN>

Assume that nothing is known about the character code transmission
capabilities of the communications channel.  If a printer's codes
setting is set to this value than PPR will not turn a job away from
that printer on the basis of the job's character codes usage.
This was the behavior of PPR prior to version 1.40.

=item B<default>

The setting is determined each time a job is printed by consulting PPR's
internal table and the printer's PPD file.  You can use the B<ppad show>
command to see the resulting setting.  Different interface programs
will have different defaults.  The default may also change according to
whether the printer's PPD file indicates that the printer supports TBCP.

=back


=item B<ppad rip> I<printer> [I<rip> I<output_language> [I<options>]]

PPR has the ability to execute external Raster Image Processors such as
Ghostscript to convert PostScript to a form usable by non-PostScript
printers.  It does so when it finds a "*pprRIP:" line in the PPD file or if
B<ppad rip> has been used to set a RIP in the printer's configuration file. 
RIP settings set with the B<ppad rip> command override those in the PPD
file.

The parameter I<rip> indicates the Raster Image Processor to use.  Normally
you will want to use a value of B<ppr-gs>.  The B<ppr-gs> RIP is a small
wrapper program for Ghostscript.

The parameter I<output_language> is the name of the printer language which
the selected Ghostscript driver will produce output in.  It is used as the
parameter to the PJL B<@PJL ENTER LANGUAGE> command.  If you will not be
using the B<pjl> or B<signal/pjl> jobbreak method, then the actual value
doesn't matter, you can set it to "x".

The option I<options> contains options for the RIP.  When using the
B<ppr-gs> RIP these should include the Ghostscript option to select the
proper output device driver.

Here is an example command which will choose B<ppr-gs> as the RIP and select
Ghostscript's driver for a HP DeskJet 500:

    $ ppad rip myprn ppr-gs pcl -sDEVICE=djet500

You should also include such other Ghostscript options as will help to configure
Ghostscript for your particular printer.  You should not include options
such as B<-sOutputFile=>, B<-q>, or B<-dSAFER> since these are supplied
automatically by B<ppr-gs>.

The RIP B<ppr-gs> also supports CUPS drivers.  Simply select the special
CUPS PPD file.  PPR understands the special lines in a CUPS PPD file that
identify the CUPS driver program that converts the output of Ghostscript to
the printer's language.  CUPS drivers will not work unless their PPD files
are selected using the B<ppad ppd> command, so there generally isn't much
need to using the B<ppad rip> command with them as PPR reads the RIP
information right out of the PPD file.  Use B<ppad show> to see what it
comes up with.  You will notice that CUPS drivers use an option called
B<cups=> which specifies the name of the CUPS driver program.

The RIP B<ppr-gs> also supports IJS drivers.  To select an IJS driver, use a
command like this one:

    $ ppad rip myprn ppr-gs escp ijs=ijsgimpprint,EPSON,escp2-c60

The B<ijs=> option takes three arguments.  They are IJS program name,
printer make, and printer model.  The set of allowable values for the last
two depends on the IJS program.  The example above is for an Epson C60
inkjet printer.  Of course, additional options may be specified:

    $ ppad rip myprn ppr-gs escp ijs=ijsgimpprint,EPSON,escp2-c60 \
        -sIjsParams=ImageType=2,MediaType=Photo,Quality=720hq2

You will have to consult the documentation for the IJS driver to find out
about available options.

If you are writing a PPD file and want to include default PPR RIP
information, you should include a line like this one.

    *pprRIP: ppr-gs pcl -sDEVICE=djet500

If you assign a PPD file with this line to a new printer but don't use the
B<ppad rip> command, it will behave as if you had entered a B<ppad rip>
command like this:

    $ ppad rip myprn ppr-gs pcl -sDEVICE=djet500

Remember though that this is a default which the system administrator will be
able to override using the B<ppad rip> command.  If he does override the
default and later wants to revert to the default, he can use this command.

    $ ppad rip myprn

Of course, if he were using a different PPD file, one with no RIP settings,
then this command would disable the external RIP and the printer will be
expected to handle PostScript itself.  To see if an external RIP is being
used, run B<ppad show> and look for a line begining with "RIP:".  If a
default from the PPD file is being used, there will be a note that so
indicates.

If you want make a RIP that you can use instead of B<ppr-gs>, you should
create a program and put it in F</usr/lib/ppr/lib/>.  It should connect
stdin, stdout, and stderr to the cooresponding file descriptors of the
PostScript interpreter and should send the print date to file descriptor 3. 
See the PPR Hackers Guide for details.

Note that while this command currently allows you to specify the path to a
copy of Ghostscript as the RIP name, you should not rely on this as it may
be removed in a future version of PPR.

=item B<ppad ppd> I<printer> I<filename>

specifies the PPD file name for the printer.  If I<filename> does not begin
with a slash, then it is looked up in the index created by B<ppr-index ppds>.  
As a last resort it is assumed to be relative to F</usr/share/ppr/ppd/>.  If
you create a printer queue with the B<ppad interface> command but do not set
the PPD file, PPR will will use a default PPD file, however, it probably
will not deal with bins properly, may print the pages in the wrong order,
and may attempt to use fonts the printer does not possess and fail to
recognized the presence of fonts it has.

Using the "ppad ppd" command has the side effect of
undoing any previous "ppad ppdopts" command.  Therefor, you should run "ppad
ppdopts" after using "ppad ppd" to set the PPD file.

=item B<ppad ppdq> I<printer>

This command uses the interface program and address previously set with the B<ppad interface>
command to contact the printer.  The printer is probed in an attempt to determine its make,
model, and other characteristis.  A search is then made of the PPD files known to
PPR.  If any of them match the make and model of the printer, then the proper
B<ppad ppd> commands to select them are printed.

Note that this command will likely fail if necessary interface options have not
yet been set with the B<ppad options> command.  It is intended to help you select
the best PPD file for a print queue that is already configured sufficiently correctly
that documents can be printed (using the default PPD file).

If you wish to probe a printer without first creating a queue for it, you should use the
B<ppad ppdlib query> command.

=item B<ppad alerts> I<printer> I<frequency> I<method> I<address>

This command specifies how PPR is to get printer error alert messages to an
operator.  This parameter is set on a per-printer basis.  For the specified
I<printer>, alerts are dispatched after every I<frequency> faults.
They are dispatched by method I<method> for which the only currently
defined value is "mail".  The message is mailed the the address
I<address>.  When a new printer queue is created with the "ppad
interface" command, a default alert setting is copied into its definition.
This default setting is set with the "ppad new alerts" command.

If the alert frequency is negative, it has special meaning.  The absolute
value is used.  When that many alerts have been posted, a message is
despatched by the method described above.  The posting of additional alerts
do not result in more messages.  When and if the printer successfully prints
a job, a message is sent indicating that the error has cleared up.

=item B<ppad frequency> I<printer> I<frequency>

Sets the alert I<frequency> for the specified I<printer> without
changing any of the other alert parameters.

=item B<ppad flags> I<printer> {B<never>, B<no>, B<yes>, B<always>} {B<never>, B<no>, B<yes>, B<always>}

Indicates whether flag pages should be printed.  The first parameter after
I<printer> is for banner pages, the second is for trailer pages.

A setting of "never" means to not print the flag page even if the user
requests it.  A setting of "no" means to not print the flag page unless the
user specifically requests it.  A setting of "yes" means to print the
flag page unless the user specifically requests that be suppressed.  A
setting of "always" means to print the flag page even if the user tries to
suppress it.

=item B<ppad outputorder> I<printer> {B<Normal>, B<Reverse>, B<PPD>}

Sets the order in which pages will be printed so as to achieve correct
collation.  If your printer stacks printed pages face up, you might want to
set outputorder to "Reverse".  If you set outputorder to "PPD", then the
setting of the output order will be determined by information in the
printer's PPD file.


=item B<ppad bins ppd> I<printer>

Sets the printer's list of bins to all the bins listed in its PPD file.  You
might create a new print queue and set its bin list
with these steps:  1) use B<ppad interface> to
create the queue, 2) use B<ppad ppd> to set the PPD file, 3) use
"ppad bins ppd" 4) use B<ppad show> to see what bin names were copied from
the PPD file, 5) use B<ppad bins delete> to remove the names of bins that
are not actually installed on the printer.

In general, a job which requires a specific type of medium will not be
printed until that medium is mounted on one or more bins.  (A medium is
mounted using the B<ppop mount> command.)

The first exception to this rule occurs when a printer has no bins in its
bin list.  In this case, it will print any job immediately without regard
to the job's media requirments.

The second exception is the bin name "AutoSelect" which has a special status.
Regardless of what medium is mounted on it, all jobs will be printed immediately
even if the proper medium is not explicitly mounted.  Pages which require a
medium which is not among those explicitly mounted will be directed to the
AutoSelect bin.


=item B<ppad bins add> I<printer> I<bin1> [I<bin2> ...]

Adds the indicated bin names to the indicated I<printer>.


=item B<ppad bins delete> I<printer> I<bin> [I<bin2> ...]

Deletes the indicated names from the indicated I<printer>.


=item B<ppad delete> I<printer>

Deletes the indicated printer.


=item B<ppad touch> I<printer>

Sends a notification to B<pprd> which causes it to re-read the configuration
file for the indicated printer.  If you edit or create a PPR printer
configuration file by hand, you should use B<ppad touch> to tell B<pprd>
that you have done so.

For example, if you keep duplicate print servers and have set up a printers
on one of them, you can then copy the configuration file to the second server
and run B<ppad touch> to tell B<pprd> about the new printer.

You can also use B<ppad touch> to duplicate a queue.  For example, if you have
a printer called "lab_1", you might use these commands to create "lab_2":

    $ cp /etc/ppr/printers/lab_1 /etc/ppr/printers/lab_2
    $ ppad touch lab_2
    $ ppad interface lab_2 tcpip lab_2.prn.myorg.org:9100

Of course, there really should be a command to do this, but there isn't.


=item B<ppad switchset> I<printer> I<switches>

Stores the indicated B<ppr> I<switches> in the printer's configuration file.
Whenever a job is sent to the I<printer>, the I<switches> will be parsed
by B<ppr> before it parses the options on its command line.

If the I<switches> are omited, then the switchset is deleted.

Switchsets can be used to set defaults for a printer which the user can
generally override with contradictory options on the B<ppr> command line.
They are also handy when the user can't select the desired options directly,
such as when the job is being submitted through Samba, L<lprsrv.8>, or
L<papd.8>.

Notice that groups do not inherit the (possibly conflicting) switchsets of
their member printers but that aliases inherit the switchset of the printer
or group to which they point.  Any switchset attached to the alias itself is 
appended to the one supplied by the printer or group to which it points.
(Note: prior to PPR version 1.53, alias did not inherit switchsets at all.)
Switchsets for groups are aliases are set with the B<ppad group switchset> 
and B<ppad alias switchset> commands respectively.

Here are some examples.  If we have a printer "myprn" and want to make duplex
the default we can do this:

	$ ppad switchset myprn --feature Duplex=DuplexNoTumble

Now, if we use the command:

	$ ppr -d myprn /etc/profile

the file will be printed in duplex mode even though we did not explicitly 
request it.  We will get the results we could get formerly by typing:

	$ ppr --feature Duplex=DuplexNoTumble -d myprn /etc/profile

But suppose we want an easy way to print in simplex mode?  We can an alias
for "myprn" which overrides the duplex setting:

	$ ppad alias what myprn_simplex myprn
	$ ppad alias switchset myprn_simplex --feature Duplex=None

Now, if we use the command:

	$ ppr -d myprn_simplex /etc/profile

then the switchset from the printer "myprn" will be applied followed by the
switchset from the alias "myprn_simplex".  The final setting will be
B<--feature Duplex=None>.

It is possible to put abusive options in switchsets.  The most obvious 

History.txt
mouse:~ppr/src/History.txt

This file contains a history of changes to PPR which occured after the
release of version 1.00.


1.00, 26 April 1995

First public release.


1.10, 27 April 1995--7 June 1995

Removed "From:" line from message passed to Sendmail by pprd when sending
a message listing the remaining printer faults.  This is so that Sendmail
will take the "real name" from the password file and put it in the "From:"
line.

Removed dotmatrix/input.epson, a GIF file which had been inadvertently
included in the distribution.

Moved the contents of include/atalk_streams.h into
interfaces/atalk_streams.c since it was not long enough to justify an
include file, nor was it of interest to any other program.

Added code to ppad to classify members of the group "ppradmin" as PPR
addministrators who can edit printer and group configurations.

Added -v switch to ppad.

Removed uid's 1-99 from the list of priveledged users.

Fixed ppad(8) so that it considers the group "ppradmin" as priveledged
rather than "ppr".

Changed ppr(1) so that it will accept the members of the group "pprprox" as
priveledged as well as the members of "ppr".  This is for security reasons
as one may not want to grant a user who has the privedge of using the -f
switch the priveledge of deleting other people's jobs.

Modified ppop(1) and pprd(8) so that user's can request to see the queue
entry for a single job.  Previously, the only options were all jobs for a
certain destination or all jobs in all queues.

Rewrote and renamed the undocumented "ppop qdebug" command to become the new
documented command "ppop details".  (This was done primarily to support
wwwppop, an experimental managment interface for PPR.  Since it was still
incomplete it was not released with this version of PPR.)

Renamed interfaces/atalk_streams.c to atalk_ali.c and
papsrv/papsrv_streams.c to papsrv_ali.c

Removed code from "make dirs" so that it doesn't create the directory
"/etc/ppr/groups.ppd".

Added sleep=N and eoftimeout=N options to tcpip interface.

Added -C _filename_ switch to ppr invokation in xppr.

Added "-default okay" to xmessage invokation in xppr.

Added the "ppop lpq" sub-command to the "ppop help" information, thus
making it a documented sub-command.

Extended automatic bin selection to multiple media types when "%%PageMedia:"
comments are present.  This has not yet been well tested.

Made "%%EndProcset" an acceptable substitute for "%%EndProcSet" due to a
reputed bug in QuickDraw GX.

Changed rule in ppr_infile.c for detecting PCL code so that documents with
only one escape would still be detected.


1.20, 8 June 1995--10 January 1996

Fixed code in ppr_infile.c so that the ppr -o switch will work even if there
is no "DefFiltOpts:" line in the printer or group configuration file.

Added '(' to list of likely second characters for HP escape codes
in ppr_infile.c

Replaced printing of errno value with strerror(errno) output in some places,
eliminated printing of errno=x from other places where both were printed.

Fixed ppr(1) so that it sets the environment variables "PATH", "IFS", and
"SHELL" so that filters and responders it invokes are less easy to subvert.

Removed setting of PATH from ~ppr/src/misc_filters/tex.perl and changed
shell name from "/usr/bin/tainperl" to "/usr/bin/perl".

Changed code which determins default responder address so that the
environment variables "USER" and "LOGNAME" are used in preference to the
current uid.

Added prompts in Configure to set the directory into which the links to PPR
binaries will be written and the path to the X-Windows binaries such as
xterm.

Modified papsrv so that it will pass the network and node number of the
client to ppr with the -r switch.  (This is for use with an experimental
AppleTalk responder.)

Modified papsrv.c, renaming cleanup() to termination_handler() and causing
it to call fatal() only the first time it is called.  Without this, name
could be left on the network if the CAP version received two terminate
signals in rapid sucession.

Added code to re-line-wrap responder messages for certain responders.
(The library function responder_width() returns the width for a responder.
Currently it is hard coded to return zero for all responders except "samba".
A return value of zero indicates that the message should not be
re-line-wrapped.)

Dropt reading of LPDEST and PRINTER by ppr(1).  (I can not really remember
why I put this feature in.  I don't think the reasons where very good.  It
may have been that I thought that the same printer would be available thru
two spooling systems so one might want to set the default destination for
both with one variable.  That now seems a little far fetched.)

Separated Solaris and Starserver sections of sysdep.h

Added PNG (Portable Network Graphics) format with magic number recognition
to ppr_infile.c.  This remains untested since I do not have any programs
which manipulate this graphics format.

In pprd.c, put sleep(1) between execl("pprdrv") and _exit(242) so that pprd
has time to record the PID before the child exits.  (If this is not done
and execl() fails immediately, a child which should have become pprdrv may
exit before the parent records its PID in the information structure for that
queue.  This result is that it will not recoginize the purpose of the child
when it receives notification of its death and will leave the queue in the
printing state indefinitely.)

Did same in pprdrv.c in code which executes interface.  Also changed
_exit(255) to _exit(242).  (Exit 242 is the new standard since PERL uses 255
as the exit code when die() is executed.)

Added automatic creation of "DefFiltOpts:" lines to ppad(8).  Manual
creation is supported by the the "ppad deffiltopts" and "ppad group
deffiltopts" command.

Modified ppop so that interactive mode restores the terminal settings when
it exits.  Also changed terminal setting code to use the POSIX method of
reading and setting terminal characteristics instead of the ioctl() method.


ppad/ppad.c
/*
** mouse:~ppr/src/ppad/ppad.c
** Copyright 1995--2006, Trinity College Computing Center.
** Written by David Chappell.
**
** This file is part of PPR.  You can redistribute it and modify it under the
** terms of the revised BSD licence (without the advertising clause) as
** described in the accompanying file LICENSE.txt.
**
** Last modified 18 May 2006.
*/

/*===========================================================================
** Administration program for PostScript page printers.  This program
** edits the media list and edits printer and group configuration
** files.
<helptopic>
	<name>misc</name>
	<desc>miscelanious command</desc>
</helptopic>
===========================================================================*/

#include "config.h"
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <pwd.h>
#include <errno.h>
#ifdef INTERNATIONAL
#include <libintl.h>
#endif
#include "gu.h"
#include "global_defines.h"
#include "util_exits.h"
#include "ppad.h"
#include "dispatch.h"
#include "dispatch_table.h"
#include "version.h"

/* misc globals */
const char myname[] = "ppad";
int machine_readable = FALSE;
int debug_level = 0;

/*
** Handle fatal errors.
** Print a message and exit.
*/
void fatal(int exitval, const char *message, ...)
	{
	va_list va;

	va_start(va,message);
	gu_utf8_fputs(_("Fatal: "), stderr);
	gu_utf8_vfprintf(stderr,message,va);
	gu_fputwc('\n', stderr);
	va_end(va);

	exit(exitval);
	} /* end of fatal() */

/*
<command helptopics="misc">
	<name><word>remind</word></name>
	<desc>send reminder e-mail about printer problems</desc>
	<args>
	</args>
</command>
*/
int command_remind(const char *argv[])
	{
	write_fifo("n\n");
	return 0;
	} /* command_remind() */

/*
** --help
*/
static void help(void)
	{
	int i;
	const char *switch_list[] =
		{
		N_("-M\tselect machine-readable output"),
		N_("--machine-readable\tsame as -M"),
		N_("--user <username>\trun as if by <username>"),
		N_("-d <n>\tset debug level to <n>"),
		N_("--debug=<n>\tsame as -d"),
		N_("--version\tprint PPR version information"),
		N_("--help\tprint this help message"),
		NULL
		};

	gu_utf8_putline(_("Valid switches:\n"));
	for(i = 0; switch_list[i]; i++)
		{
		const char *p = gettext(switch_list[i]);
		int to_tab = strcspn(p, "\t");
		gu_utf8_printf("    %-20.*s %s\n",
			to_tab, p,
		   	p[to_tab] == '\t' ? &p[to_tab + 1] : ""
			);
		}

	gu_putwc('\n');

	{
	const char *args[] = {"help", NULL};
	dispatch(myname, args);
	}

	gu_putwc('\n');
	gu_utf8_printf(
		_(	"The %s manpage may be viewed by entering this command at a shell prompt:\n"
			"    ppdoc %s\n"
			),
	   	"ppad(1)",
	   	"ppad"
		);
	} /* help() */

static const char *option_chars = "Md:";
static const struct gu_getopt_opt option_words[] =
		{
		{"machine-readable", 'M', FALSE},
		{"debug", 'd', TRUE},
		{"help", 1000, FALSE},
		{"version", 1001, FALSE},
		{"user", 1002, TRUE},
		{(char*)NULL, 0, FALSE}
		} ;

/*
** Main function.
**
** This depends on the argv[] array being terminated by
** a NULL character pointer.
*/
int main(int argc, char *argv[])
	{
	int optchar;
	struct gu_getopt_state getopt_state;

	/* Initialize internation messages library. */
	gu_locale_init(argc, argv, PACKAGE, LOCALEDIR);

	/* So config files come out with desired mode. */
	umask(PPR_UMASK);

	/* Figure out the user's name and make it the initial value for --user. */
	{
	struct passwd *pw;
	uid_t uid = getuid();
	if(!(pw = getpwuid(uid)))
		{
		gu_utf8_fprintf(stderr, "%s: getpwuid(%ld) failed, errno=%d (%s)\n", myname, (long)uid, errno, gu_strerror(errno));
		return EXIT_INTERNAL;
		}
	dispatch_set_user(NULL, pw->pw_name);
	}

	/*
	** Parse any dash options which appear before the subcommand.
	*/
	gu_getopt_init(&getopt_state, argc, argv, option_chars, option_words);
	while((optchar = ppr_getopt(&getopt_state)) != -1)
		{
		switch(optchar)
			{
			case 'M':							/* machine readable */
				machine_readable = TRUE;
				/* Send error messages to stdout. */
				/*stderr = stdout;*/
				dup2(1,2);
				break;

			case 'd':							/* debug */
				if(strspn(getopt_state.optarg, "0123456789") != strlen(getopt_state.optarg))
					{
					gu_utf8_fprintf(stderr, _("%s: debug level must be a positive integer\n"), myname);
					exit(EXIT_SYNTAX);
					}
				debug_level = atoi(getopt_state.optarg);
				break;

			case 1000:							/* --help */
				help();
				exit(EXIT_OK);

			case 1001:							/* --version */
				if(machine_readable)
					{
					gu_utf8_putline(SHORT_VERSION);
					}
				else
					{
					gu_utf8_putline(VERSION);
					gu_utf8_putline(COPYRIGHT);
					gu_utf8_putline(AUTHOR);
					}
				exit(EXIT_OK);

			case 1002:							/* --user */
				if(dispatch_set_user("ppad", getopt_state.optarg) == -1)
					{
					gu_utf8_fprintf(stderr, _("%s: you are not allowed to use --user\n"), myname);
					exit(EXIT_DENIED);
					}
				break;

			default:
				gu_getopt_default(myname, optchar, &getopt_state, stderr);
				exit(EXIT_SYNTAX);
				break;
			}
		}

	/*
	** If there is a command, dispatch it, otherwise
	** invoke interactive mode.
	*/
	if(getopt_state.optind < argc)
		return dispatch(myname, (const char **)&argv[getopt_state.optind]);
	else
		return dispatch_interactive(myname, _("PPAD, Page Printer Administrator's utility"), "ppad>", machine_readable);

	} /* end of main() */

/* end of file */

